Kafka upgrade (#2871) (#2803)

cicharka opened the pull request:
* Kafka: Upgrade to 2.8.1 version

* Kafka: Refactor installation and upgrade tasks

* Kafka: move upgrade tasks to Kafka role

* Zookeeper: refactor installation and upgrade

* Zookeeper: move upgrade tasks to Zookeeper role

* Kafka: Refactor: rename 'kafka_var' setting

* Kafka-exporter: verify if service definition needs to be updated even if exporter itself was not updated

* Kafka-exporter: remove kafka.version from configuration

cicharka committed:
Zookeeper: move upgrade tasks to Zookeeper role

* upgrade playbook updated in order to use upgrade
  zookeeper tasks from a role

cicharka committed:
Zookeeper: refactor installation and upgrade

cicharka committed:
Zookeeper: small fixes

Signed-off-by: cicharka <arkadiusz.cichon@outlook.com>

cicharka committed:
Kafka: move upgrade tasks to Kafka role

cicharka committed:
Kafka: Refactor installation and upgrade tasks

cicharka committed:
Kafka: upgrade to version 2.8.1

cicharka committed:
Remove kafka_var #2803

cicharka committed:
Changelog updated

przemyslavic commented:
Don't we want to update ZooKeeper to v3.5.9? https://zookeeper.apache.org/doc/r3.5.9/releasenotes.html
It comes with Kafka 2.8.0.

cicharka commented:
> Don't we want to update ZooKeeper to v3.5.9? https://zookeeper.apache.org/doc/r3.5.9/releasenotes.html It comes with Kafka 2.8.0.

I think that I didn't take it into account because:
* documentation of Ecosystem for Kafka 2.8.1 says nothing about Zookeeper upgrade -> it only mention that in Kafka 2.5.0 ZooKeeper has been upgraded to 3.5.7. Additionally I have not checked Kafka archive in order to investigate Zookeeper version which comes with it (didn't know about it to be honest)
* I checked that Kafka doesn't specify Zookeeper version required, and as we are installing it as standalone component and I have not found any issues that states to upgrade it, I just skipped it

However, since you asked about it, and even found that it comes with Kafka, we can think about 2 options:
1. We are going to remove ZooKeeper role since it is only used as prerequisite for Kafka (for my understanding, if it's not true please correct me) and we are going to use Zookeeper that comes with Kafka (additionally Kafka states that they already removed ZooKeeper from Kafka, but still in development stage)
2. We are going only to upgrade ZooKeeper version -> ZooKeeper role will stay, only defaults and binary changed needed.

I don't know about any other ZooKeeper usecases but if its only Kafka's prerequisite I would vote for role removal and move to use ZooKeeper that comes with Kafka. (I would also do it in another PR to do not overwhelm reviewers :) )
Sorry for long post!

cicharka committed:
Kafka: fix upgrade and update components

rafzei commented on ansible/playbooks/roles/kafka/tasks/common/download_and_unpack_binary.yml:
```diff
@@ -0,0 +1,22 @@
+---
+
```
```suggestion

```

rafzei commented on ansible/playbooks/roles/kafka/tasks/common/stop.yml:
```diff
@@ -0,0 +1,6 @@
+---
+
```
```suggestion

```

rafzei commented on ansible/playbooks/roles/kafka/tasks/generate-certificates.yml:
```diff
@@ -1,55 +1,55 @@
-    group: "{{ specification.kafka_var.group }}"
+    owner: "{{ specification.user }}"
+    group: "{{ specification.group }}"
     mode: "0755"
```
Some time ago we agreed to use symbolic modes
```suggestion
    mode: u=rwx,go=rx
```
BTW do we need that wide permissions? (please check also in other places)

rafzei commented on ansible/playbooks/roles/kafka/tasks/generate-certificates.yml:
```diff
@@ -1,55 +1,55 @@
-         -alias localhost -validity {{ specification.kafka_var.security.ssl.server.cert_validity }} -genkey -keyalg RSA \
-         -noprompt -storepass {{ specification.kafka_var.security.ssl.server.passwords.keystore }} \
-         -keypass {{ specification.kafka_var.security.ssl.server.passwords.key }} \
+  shell: keytool -keystore {{ specification.security.ssl.server.keystore_location }} \
```
I would use `command` here and in other places w/o chars like <,>, |  and &

cicharka replied:
That was the only place where we are not using any shell operators, so I only changed it in this line.

rafzei commented on ansible/playbooks/roles/kafka/tasks/metrics.yml:
```diff
@@ -2,17 +2,17 @@
-    group: "{{ specification.kafka_var.group }}"
+    owner: "{{ specification.user }}"
+    group: "{{ specification.group }}"
     mode: 0644
```
```suggestion
    mode: u=rx,go=r
```

rafzei commented on ansible/playbooks/roles/kafka/tasks/setup-kafka.yml:
```diff
@@ -24,48 +24,20 @@
   copy:
     content: "export PATH=$PATH:/opt/kafka/bin"
     dest: "/etc/profile.d/kafka_path.sh"
     mode: 0755
```
```suggestion
    mode: u=rwx,go=rx
```

rafzei commented on ansible/playbooks/roles/kafka/tasks/setup-kafka.yml:
```diff
@@ -78,59 +50,60 @@
-    group: "{{ specification.kafka_var.group }}"
+    owner: "{{ specification.user }}"
+    group: "{{ specification.group }}"
     mode: 0755
```
```suggestion
    mode: u=rwx,go=rx
```

rafzei commented on ansible/playbooks/roles/kafka/tasks/setup-kafka.yml:
```diff
@@ -78,59 +50,60 @@
-    group: "{{ specification.kafka_var.group }}"
+    owner: "{{ specification.user }}"
+    group: "{{ specification.group }}"
     mode: 0755
```
```suggestion
    mode: u=rwx,go=rx
```

rafzei commented on ansible/playbooks/roles/kafka/tasks/setup-kafka.yml:
```diff
@@ -78,59 +50,60 @@
+    path: "{{ specification.conf_dir }}/log4j.properties"
+    owner: "{{ specification.user }}"
+    group: "{{ specification.group }}"
     mode: 0644
```
```suggestion
    mode: u=rw,go=r
```

rafzei commented on ansible/playbooks/roles/kafka/tasks/setup-kafka.yml:
```diff
@@ -78,59 +50,60 @@
+    owner: "{{ specification.user }}"
+    group: "{{ specification.group }}"
     # Was 0640
     mode: 0644
```
```suggestion
    mode: u=rw,go=r
```

rafzei commented on ansible/playbooks/roles/kafka/tasks/generate-certificates.yml:
```diff
@@ -69,71 +69,71 @@
 - name: Check if CA certificate is already imported
-  shell: keytool -list -v -keystore {{ specification.kafka_var.security.ssl.server.keystore_location }} \
-         -storepass {{ specification.kafka_var.security.ssl.server.passwords.keystore }} \
+  shell: keytool -list -v -keystore {{ specification.security.ssl.server.keystore_location }} \
```
I would add `set -o pipefail` here

cicharka replied:
done.

rafzei commented on ansible/playbooks/roles/zookeeper/tasks/main.yml:
```diff
@@ -112,7 +70,11 @@
+    path: /opt/zookeeper
+    state: link
+    src: "{{ zookeeper_install_dir }}"
+  # force: ?
```
:smiley_cat:

cicharka replied:
yeah :( fixed now!

cicharka committed:
kafka and zookeeper: fixes after review

erzetpe commented on ansible/playbooks/roles/kafka/tasks/common/download_and_unpack_binary.yml:
```diff
@@ -0,0 +1,21 @@
+    path: "{{ kafka_install_dir }}"
+    state: directory
+    owner: kafka
+    group: kafka
```
This probably should be taken from vars?

cicharka replied:
Taken from specification as we let user configure that.

erzetpe commented on ansible/playbooks/roles/kafka/tasks/upgrade/install-upgrade.yml:
```diff
@@ -0,0 +1,34 @@
+    src: "{{ kafka_install_dir}}"
+    force: yes
+
+- name: Remove previous version binaries
+  file:
+    path: /opt/kafka_{{ scala_version }}-{{ old_kafka_version }}
+    state: absent
```
We should confirm that we have something about backup or put some flag to left the people decision if they want to remove old binaries.

cicharka replied:
Added note in Upgrade docs

cicharka committed:
Kafka: use specification to set directory permissions

rafzei reviewed (dismissed):
LGTM

atsikham commented on ansible/playbooks/roles/kafka/tasks/common/download_and_unpack_binary.yml:
```diff
@@ -0,0 +1,21 @@
+
+- name: Uncompress the Kafka tar
+  unarchive:
+    remote_src: yes
```
We agreed some time before to use `true/false` values, please change in each affected place.

cicharka replied:
Ok, I will review other places as well

atsikham commented on ansible/playbooks/roles/kafka/tasks/common/stop.yml:
```diff
@@ -0,0 +1,5 @@
+---
```
Is it needed to be stored in a separate file? Same for start.yml.

cicharka replied:
start.yml was already in separate file so I only [moved it to common](https://github.com/epiphany-platform/epiphany/pull/2990/files/13f1487d2d1082019b1e57cc51725dd70d3176f9#diff-42a0c302e962e0a498efae3dd5167585a4703e00bc935533284ff0f58408feff), and I did the same for stop.yml just to be consistent with let's call it, 'common tools'.
I know that it is not the biggest improvement...

atsikham commented on ansible/playbooks/roles/kafka/tasks/generate-certificates.yml:
```diff
@@ -1,55 +1,55 @@
 - name: Check if signing certificate exists
   stat:
-    path: "{{ specification.kafka_var.security.ssl.server.keystore_location | dirname }}/ca-cert"
+    path: "{{ specification.security.ssl.server.keystore_location | dirname }}/ca-cert"
```
Add `get_*` parameters with `false` value where it's possible.
https://docs.ansible.com/ansible/latest/collections/ansible/builtin/stat_module.html

atsikham commented on ansible/playbooks/roles/kafka/tasks/generate-certificates.yml:
```diff
@@ -1,55 +1,55 @@
-  shell: openssl req -new -x509 -keyout {{ specification.kafka_var.security.ssl.server.keystore_location | dirname }}/ca-key \
-         -out {{ specification.kafka_var.security.ssl.server.keystore_location | dirname }}/ca-cert \
-         -days {{ specification.kafka_var.security.ssl.server.cert_validity }} \
+  shell: openssl req -new -x509 -keyout {{ specification.security.ssl.server.keystore_location | dirname }}/ca-key \
```
Change `shell` module to `command` where it's possible, like it's done above.

cicharka replied:
Already corrected some of them based on @rafzei comments, but you are right, there are still places that can be improved.

atsikham commented on ansible/playbooks/roles/kafka/tasks/generate-certificates.yml:
```diff
@@ -1,55 +1,55 @@
   delegate_to: localhost
 
 - name: Fetching files
   fetch:
```
May be changed to `slurp` not to store local files.

atsikham commented on ansible/playbooks/roles/kafka/tasks/metrics.yml:
```diff
@@ -52,7 +52,7 @@
+    mode: u=rx,go=r
   delegate_to: "{{ item }}"
   notify: restart prometheus
   with_inventory_hostnames:
```
I suggest to use `loop` to follow the same standard as we have in other places.

atsikham commented on ansible/playbooks/roles/kafka/tasks/upgrade/main.yml:
```diff
@@ -1,25 +1,16 @@
-
 - name: Include pre-flight checks
-  include_tasks: kafka/preflight-check.yml
+  include_tasks: upgrade/preflight-check.yml
```
In `postgresql` role we have the `preflight` directory, what do you think about doing the same here?

cicharka replied:
I checked it now, cool idea -> but Kafka doesn't have any preflight checks file for apply.

But it doesn't mean we cannot adapt that idea in Kafka:
Task `- name: Check if jmx exporter is available`
from `tasks/main.yml` can be extracted into newly created `tasks/prefilight/apply.yml`, and `upgrade/preflight-check.yml` can be moved into `tasks/preflight/upgrade.yml`.

As additional note, I would encourage ourselves to note that approach and let other developers know about that structure.

atsikham commented on ansible/playbooks/roles/kafka/tasks/upgrade/main.yml:
```diff
@@ -1,25 +1,16 @@
   register: result
 
-- name: Set common facts
+- name: Set old Kafka version fact
```
How about usage of the fact's name here? `old_kafka_version `

atsikham reviewed (changes requested):
This PR also can be the place for this task - https://github.com/epiphany-platform/epiphany/issues/2235. As I see, openjdk 8 is used only for Kafka and Zookeeper.

cicharka committed:
* refactor more tasks

cicharka committed:
* change 'yes' to 'true' in tasks

cicharka committed:
fix permission

rafzei commented:
@cicharka One more thing, could you please run ansible-lint task in devcontainer on Kafka playbook and fix the issues?

cicharka commented:
/azp run

cicharka committed:
fix ansible-lint error for kafka role

cicharka committed:
fix ansible-lint errors for zookeeper role

cicharka commented:
/azp run

przemyslavic commented:
/azp run

przemyslavic commented:
:x: It fails when there is no monitoring component enabled:
```
2022-03-23T11:45:38.0521213Z [38;21m11:45:38 INFO cli.src.ansible.AnsibleCommand - TASK [kafka : delegated | create prometheus system group] **********************[0m
2022-03-23T11:45:38.0781789Z [31;21m11:45:38 ERROR cli.src.ansible.AnsibleCommand - fatal: [ci-kfkawsrhelflannel-kafka-vm-0 -> {{ node }}]: FAILED! => {"msg": "'dict object' has no attribute 'prometheus'"}[0m
2022-03-23T11:45:38.1062569Z [31;21m11:45:38 ERROR cli.src.ansible.AnsibleCommand - fatal: [ci-kfkawsrhelflannel-kafka-vm-1 -> {{ node }}]: FAILED! => {"msg": "'dict object' has no attribute 'prometheus'"}[0m
2022-03-23T11:45:38.1160160Z [31;21m11:45:38 ERROR cli.src.ansible.AnsibleCommand - fatal: [ci-kfkawsrhelflannel-kafka-vm-2 -> {{ node }}]: FAILED! => {"msg": "'dict object' has no attribute 'prometheus'"}[0m
2022-03-23T11:45:38.1162720Z [38;21m11:45:38 INFO cli.src.ansible.AnsibleCommand - [0m
```

cicharka committed:
Add verification for prometheus groups existence

przemyslavic commented:
/azp run

cicharka committed:
kafka: adjust kafka.version in kafka exporter schema

przemyslavic commented:
/azp run

przemyslavic commented:
/azp run

cicharka committed:
ensure that kafka_exporter definition is up to date
with currently used kafka version

cicharka committed:
remove kafka.version configuration flag from schema

* We have specified version of kafka used in epiphany, therefore it
  should not be possible for user to adjust its version in
  kafka_exporter flags used in service definition

cicharka committed:
* adjust service definition verification

przemyslavic commented:
/azp run

przemyslavic commented:
/azp run
